### DISCUSSION TITLE
Confused by the `upsert`.

### DISCUSSION THREAD

**Discussion by teenjuna** [Brendonovich/prisma-client-rust#111] [Q&A] [answered]
Hey! I'm really confused by `upsert`. It looks like I should use it when I want to save some record, by not sure if it already exists or not. Usually this happens when I have some struct with optional `id` field (e.g. `id: Option<i32>`). If `id.is_some()`, I want to update, otherwise - create. Seems like the ideal case for `upsert`, but it requires `id` to be specified! That's... confusing. Though I never used the original Prisma client and don't know if it's meant to be like that.

Also, the syntax for arguments is a bit confusing, too. Let's look at the example provided in docs:
```rust
let post: post::Data = client
    .post()
    .upsert( // First argument is a unique filter
        post::id::equals("upsert".to_string()),
        ( // Second argument is a tuple with the same values
          // as an equivalent call to create()
            true,
            "title".to_string(),
            "upsert".to_string(),
            vec![]
        ),
        vec![ // Final argument is a vec of updates
            post::content::set(Some("new content".to_string())),
            post::views::increment(1)
        ]
    )
    .exec()
    .await
    .unwrap();
```
I have two questions:
1. Why is the `content::set(...)` in the vec of updates and not in the vec in the tuple, like this?
```rust
let post: post::Data = client
    .post()
    .upsert( // First argument is a unique filter
        post::id::equals("upsert".to_string()),
        ( // Second argument is a tuple with the same values
          // as an equivalent call to create()
            true,
            "title".to_string(),
            "upsert".to_string(),
            vec![
                post::content::set(Some("new content".to_string())),
            ]
        ),
        vec![ // Final argument is a vec of updates
            post::views::increment(1)
        ]
    )
    .exec()
    .await
    .unwrap();
```
2. Why is the second argument must be in a tuple? Maybe the idea was in a mental separation, or is it `clippy`'s `too_many_arguments` lint? If the former is true, I want to say that if feels unnecessary because the vec of updates is always the last argument. If the latter is true, I would suggest just attaching `#[allow(clippy::too_many_arguments)]` to the method in codegen part.

**Comment by Brendonovich** [comment, marked as answer]
While admittedly the example in the docs may not be the easiest to understand, the structure is correct. To address your points individually:
- Why provide the ID? Upsert exists to create a _record_ if it doesn't exist, and if the record _does_ exist then update it with the provided values. Prisma needs a unique field to determine whether the record exists or not and decide whether to create or update.

1. It could be, but `content` in the create section is already specified as `"upsert".to_string()`. The example is meant to demonstrate assigning different values depending on whether the record exists or not.
2. The structure of the arguments is `upsert(where, create, update)`. `where` is the same argument as accepted as a call to `find_unique`, `create` is the same arguments accepted by a call to `create`, and `update` is the same arguments as accepted by a call to `update`. While I could flatten the `create` arguments into the arguments of `upsert`, I prefer having a distinct separation between which data is for creation and which is for updating. Either extra arguments to `upsert` or that tuple are required, as I have to be able to enforce that values for required fields are provided by the user.

**Reply by teenjuna** [reply]
Considering the second point: so if I just left the update vec empty, this operation won't modify existing record at all?

Considering the ID, is it possible to integrate `Option` in the future? Something like this:
1. ID is `None` -> upsert = create
2. ID is `Some(id)`, but `id` doesn't exist -> upsert = create
3. ID is `Some(id)` and `id` exists -> upsert = update

**Reply by Brendonovich** [reply]
> Considering the second point: so if I just left the update vec empty, this operation won't modify existing record at all?

Correct

> Considering the ID, is it possible to integrate `Option` in the future? Something like this:
> 
> 1. ID is `None` -> upsert = create
> 2. ID is `Some(id)`, but `id` doesn't exist -> upsert = create
> 3. ID is `Some(id)` and `id` exists -> upsert = update

Most likely no. Prisma requires that the `where` argument is unique, which does not include `None` values. If you think about it, that doesn't even make sense. If `id` is `None` and there are multiple records with a `NULL` id then how should Prisma choose which record to upsert?

**Reply by teenjuna** [reply]
Fair enough! Thank for your answer :)
